test/test_foreach.py

```diff
@@ -599,11 +604,42 @@ def test_complex_scalarlist(self, device, dtype):
                             foreach_bin_op(tensors, scalar)
                     return
 
-                with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
+                if self.device_type == 'cpu':
                     res = foreach_bin_op(tensors, scalars)
+                    self.assertEqual(res, expected)
 
-                with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
-                    foreach_bin_op_(tensors, scalars)
+                    if dtype not in [torch.complex64, torch.complex128]:
+                        with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                            foreach_bin_op_(tensors, scalars)
+                    else:
+                        self.assertEqual(res, expected)
+                else:
+                    # float types
+                    if dtype in torch.testing.get_all_fp_dtypes(include_half=True, include_bfloat16=True):
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op_(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op(tensors, scalars)
+                        return
+
+                    # complex types
+                    if dtype in [torch.complex64, torch.complex128]:
+                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                            foreach_bin_op(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                            foreach_bin_op_(tensors, scalars)
+                        return
+
+                    # int types
+                    if dtype in torch.testing.integral_types():
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op_(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op(tensors, scalars)
+                        return
```

Comment:
Some things I am concerned about:
- Why does the complex behavior differ on CPU/CUDA? Do we want those to be ultimately the same?
- The integral cuda and the float cuda logic the same to me, should we just combine them?

Reply:
Please, see the next PR in this stack. I've rewrote the tests to have a separate test per operation. For example, this is how this tests looks [there](https://github.com/pytorch/pytorch/blob/87ce1f9f10b3f694987ff4aea07b12f08effb67c/test/test_foreach.py#L1464).

Much cleaner and easier to follow.